src/core_plugins/kibana/public/home/components/sample_data_set_cards.js

```diff
@@ -0,0 +1,191 @@
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import _ from 'lodash';
+import React from 'react';
+import PropTypes from 'prop-types';
+import {
+  EuiFlexGrid,
+  EuiFlexItem,
+} from '@elastic/eui';
+
+import { SampleDataSetCard } from './sample_data_set_card';
+
+import { toastNotifications } from 'ui/notify';
+
+import {
+  listSampleDataSets,
+  installSampleDataSet,
+  uninstallSampleDataSet
+} from '../sample_data_sets';
+
+export class SampleDataSetCards extends React.Component {
+
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      sampleDataSets: [],
+      processingStatus: {},
+    };
+  }
+
+  componentWillUnmount() {
+    this._isMounted = false;
+  }
+
+  async componentDidMount() {
+    this._isMounted = true;
+
+    this.loadSampleDataSets();
+  }
+
+  loadSampleDataSets = async () => {
+    let sampleDataSets;
+    try {
+      sampleDataSets = await listSampleDataSets();
+    } catch (fetchError) {
+      toastNotifications.addDanger({
+        title: `Unable to load sample data sets list`,
+        text: `${fetchError.message}`,
+      });
+      sampleDataSets = [];
+    }
+
+    if (!this._isMounted) {
+      return;
+    }
+
+    this.setState({
+      sampleDataSets: sampleDataSets
+        .sort((a, b) => {
+          return a.name.toLowerCase().localeCompare(b.name.toLowerCase());
+        }),
+      processingStatus: {},
+    });
+  }
+
+  install = async (id) => {
+    const {
+      getConfig,
+      setConfig,
+      clearIndexPatternsCache,
+    } = this.props;
+
+    const targetSampleDataSet = this.state.sampleDataSets.find((sampleDataSet) => {
+      return sampleDataSet.id === id;
+    });
+
+    this.setState({
+      processingStatus: { ...this.state.processingStatus, [id]: true }
+    });
+
+    try {
+      await installSampleDataSet(id, targetSampleDataSet.defaultIndex, getConfig, setConfig, clearIndexPatternsCache);
+    } catch (fetchError) {
+      if (this._isMounted) {
+        this.setState({
+          processingStatus: { ...this.state.processingStatus, [id]: false }
+        });
+      }
+      toastNotifications.addDanger({
+        title: `Unable to install sample data set: ${targetSampleDataSet.name}`,
+        text: `${fetchError.message}`,
+      });
+      return;
+    }
+
+    await this.loadSampleDataSets();
```

Comment:
This is hard to verify with only one actual sample data set available, but I think an error could appear if the client hit add consecutively on multiple data sets.  It appears that the first response from `installSampleDataSet` will call `loadSampleDataSets` which clears out all processingStatus states.

Reply:
should the local processingStatus be preserved?

Reply:
I just suspect a bug will arise when we add more sample data sets, where if you pressed "Add", "Add", "Add", all three states go to "true", so then you'd see "Adding", "Adding", "Adding".  When the first one returns, it'll set all three to false.  Reloading the data might show that the one was successfully added, but the others, may not have finished processing, so I think now you'd see "Remove", "Add", "Add".  Even though the other two are still in the process of being added.